// @axe-core/watcher/dist/cypress/cypress.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cypressConfig = exports.cypressRunOptionsFlag = exports.cypressAxeCoreSettingsFlag = exports.cypressConfigAutoAnalyzeFlag = exports.cypressConfigRunFlag = exports.configTimeout = void 0;
const sendResultsToServer_1 = __importDefault(require("../sendResultsToServer"));
const EventForwarder_1 = __importDefault(require("../EventForwarder"));
const writeVariables_1 = __importDefault(require("../utils/writeVariables"));
const writeExtensionManifest_1 = __importDefault(require("../utils/writeExtensionManifest"));
const headlessNotSupportedError_1 = require("../utils/headlessNotSupportedError");
const mergeChromeArgs_1 = __importDefault(require("../utils/mergeChromeArgs"));
const createDebugger_1 = __importDefault(require("../createDebugger"));
const readVariables_1 = __importDefault(require("../utils/readVariables"));
const checkUserArgs_1 = require("../utils/checkUserArgs");
const debugLogger = (0, createDebugger_1.default)('CypressController');
exports.configTimeout = '__AXE_WATCHER_TIMEOUT_CONFIG';
exports.cypressConfigRunFlag = '__AXE_WATCHER_CYPRESS_CONFIG_RUN';
exports.cypressConfigAutoAnalyzeFlag = '__AXE_WATCHER_AUTO_ANALYZE';
exports.cypressAxeCoreSettingsFlag = '__AXE_WATCHER_AXE_CORE_SETTINGS';
exports.cypressRunOptionsFlag = '__AXE_WATCHER_RUN_OPTIONS';
const cypressConfig = (config) => {
    const { axe, ...userConfig } = config;
    const { DEBUG } = process.env;
    const hasWatcherDebug = DEBUG && DEBUG.includes('axe-watcher');
    (0, writeVariables_1.default)({
        ...axe,
        cypress: true
    });
    (0, writeExtensionManifest_1.default)({
        all_frames: true,
        exclude_globs: [
            '*/__/#/specs/runner?file=*',
            '*/__cypress/iframes/*'
        ]
    });
    const variables = (0, readVariables_1.default)();
    if (userConfig.e2e?.env) {
        userConfig.e2e.env = Object.assign({}, userConfig.e2e.env, {
            [exports.cypressConfigRunFlag]: true,
            [exports.configTimeout]: variables.timeout,
            [exports.cypressAxeCoreSettingsFlag]: variables.axe_core_settings,
            [exports.cypressRunOptionsFlag]: variables.run_options
        });
    }
    else {
        userConfig.env = Object.assign({}, userConfig.env, {
            [exports.cypressConfigRunFlag]: true,
            [exports.configTimeout]: variables.timeout,
            [exports.cypressAxeCoreSettingsFlag]: variables.axe_core_settings,
            [exports.cypressRunOptionsFlag]: variables.run_options
        });
    }
    if (hasWatcherDebug) {
        userConfig.env = Object.assign({}, userConfig.env, {
            __AXE_WATCHER_DEBUG: hasWatcherDebug
        });
    }
    userConfig.env = Object.assign({}, userConfig.env, {
        [exports.cypressConfigAutoAnalyzeFlag]: axe.autoAnalyze === undefined ? true : axe.autoAnalyze
    });
    return {
        ...userConfig,
        e2e: {
            ...userConfig.e2e,
            async setupNodeEvents(cypressOn, ...args) {
                const eventForwarder = new EventForwarder_1.default();
                const on = eventForwarder.on;
                const userNodeEventConfig = await userConfig.e2e?.setupNodeEvents?.(on, ...args);
                if (userNodeEventConfig) {
                    userNodeEventConfig.env = Object.assign({}, userNodeEventConfig.env, {
                        [exports.cypressConfigRunFlag]: true
                    });
                }
                on('task', {
                    __uploadAxeWatcherResults: async ({ results, skipped_url }) => {
                        if (!results?.length && !skipped_url) {
                            return null;
                        }
                        await (0, sendResultsToServer_1.default)({
                            skipped_url,
                            results,
                            debugLogger
                        });
                        return null;
                    },
                    __debugAxeWatcher: (message) => {
                        debugLogger(message);
                        return null;
                    }
                });
                on('before:browser:launch', (browser, launchOptions) => {
                    if (browser.name.startsWith('firefox') ||
                        browser.name.startsWith('electron')) {
                        throw new Error(`The @axe-core/watcher package only supports running in Chrome, but Cypress is configured with browser "${browser.name}". Please refer to our System Requirements for more information: https://docs.deque.com/developer-hub/2/en/dh-system-requirements`);
                    }
                    if (browser.isHeadless && launchOptions.args.includes('--headless')) {
                        throw new headlessNotSupportedError_1.HeadlessNotSupportedError('In Cypress, "--headless=new" became the default headless mode in Cypress version 12.15.0. Ensure your Cypress version is up to date and that you are not using a "before:browser:launch" configuration that adds "--headless" to launchOptions.args.');
                    }
                    const existingArgs = launchOptions.args;
                    (0, checkUserArgs_1.checkUserArgs)(existingArgs);
                    launchOptions.args = (0, mergeChromeArgs_1.default)(existingArgs, {
                        disableOtherExtensions: false
                    });
                    return launchOptions;
                });
                eventForwarder.forward(cypressOn);
                return userNodeEventConfig;
            }
        }
    };
};
exports.cypressConfig = cypressConfig;
